File: go/parsers/cancel.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Cancel struct {
	gen *gen.Cancel
}

func NewCancelFields(g *gen.Cancel) messages.CancelFields {
	return &Cancel{gen: g}
}

func (c *Cancel) RequestID() uint64 {
	return c.gen.RequestID()
}

func (c *Cancel) Options() map[string]any {
	return map[string]any{}
}

func CancelToCapnproto(m *messages.Cancel) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	c, err := gen.NewRootCall(seg)
	if err != nil {
		return nil, err
	}

	c.SetRequestID(m.RequestID())

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeCancel, data, nil), nil
}

func CapnprotoToCancel(data []byte) (*messages.Cancel, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	cancel, err := gen.ReadRootCancel(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewCancelWithFields(NewCancelFields(&cancel)), nil
}

File: python/wamp_msgs_capnp/parsers/cancel.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.cancel import Cancel, ICancelFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "cancel.capnp")
cancel_capnp = capnp.load(str(module_file))


class CancelFields(ICancelFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def options(self) -> dict[str, Any]:
        return {}


def cancel_to_capnproto(c: Cancel) -> bytes:
    cancel = cancel_capnp.Cancel.new_message()
    cancel.requestID = c.request_id

    packed_data = cancel.to_bytes_packed()

    return helpers.prepend_header(Cancel.TYPE, packed_data, b"")


def capnproto_to_cancel(data: bytes) -> Cancel:
    message_data, _ = helpers.extract_message(data)
    cancel_obj = cancel_capnp.Cancel.from_bytes_packed(message_data)

    return Cancel(CancelFields(cancel_obj))
